### PR TITLE
feat: awarded course

### PR DESCRIPTION
**Pull request by Michellah**

### CODE REVIEW

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  required,
+  SelectInput,
+  SimpleForm,
+  useGetList,
+} from "react-admin";
+import {useParams} from "react-router-dom";
+
+interface AwardedCoursesCreateProps {
+  toggleShowCreate: () => void;
+}
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx:14]: skip a line after this

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+export const AwardedCoursesCreate: React.FC<AwardedCoursesCreateProps> = ({
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: import FC from react

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  toggleShowCreate,
+}) => {
+  const params = useParams();
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx:22]: order the variables; Hook before simple variable

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  const courseId = params.id;
+  const {data: teachers} = useGetList("teachers");
+  const {data: groups} = useGetList("groups");
+  const teacherChoices = teachers
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: skip a line after hooks

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    ? teachers.map((teacher) => ({
+        id: teacher.id,
+        name: `${teacher.first_name} ${teacher.last_name}`,
+      }))
+    : [];
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: skip a line

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  const groupChoices = groups
+    ? groups.map((group) => ({
+        id: group.id,
+        ref: group.ref,
+      }))
+    : [];
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: skip a line

File: src/operations/awardedCourses/AwardedCoursesList.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import {TextField} from "react-admin";
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/AwardedCoursesList.tsx]: reorder imports

File: src/operations/awardedCourses/TeacherCourseList.tsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import {TextField} from "react-admin";
+import {MenuBook as BookIcon} from "@mui/icons-material";
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/TeacherCourseList.tsx:2]: reorder imports

File: src/operations/awardedCourses/TeacherCourseList.tsx [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import {HaList} from "@/ui/haList";
+import authProvider from "@/providers/authProvider";
+import {useRole} from "@/security/hooks";
+
+export const TeacherCourseList = () => {
+  const {isTeacher} = useRole();
+  const teacherId = isTeacher() ? authProvider.getCachedWhoami().id : null;
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/TeacherCourseList.tsx]: skip a line

File: src/operations/awardedCourses/TeacherCourseList.tsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  return (
+    <HaList
+      icon={<BookIcon />}
+      title="Cours"
+      resource="awarded-courses"
+      actions={""}
```
Comment (NyAndoMayah) [inline src/operations/awardedCourses/TeacherCourseList.tsx]: just add false or idk

File: src/operations/course/CourseList.tsx [hunk 11]
```diff
@@ -1,33 +1,32 @@
-import {useState} from "react";
-import {TextField, useNotify} from "react-admin";
+import {ShowButton, TextField, useNotify} from "react-admin";
 import {Box} from "@mui/material";
 import {Book, Add as AddIcon} from "@mui/icons-material";
+import {CourseListFilter} from "@/operations/course/components";
```
Comment (NyAndoMayah) [inline src/operations/course/CourseList.tsx:5]: reorder imports

File: src/operations/course/CourseList.tsx [hunk 12]
```diff
@@ -1,33 +1,32 @@
...
 
 export function CourseList() {
   const [showCreate, _set, toggleShowCreate] = useToggle();
-  const [selectedCourseId, setSelectedCourseId] = useState(null);
-
+  const {isTeacher} = useRole();
   const notify = useNotify();
 
-  return (
+  return isTeacher() ? (
```
Comment (NyAndoMayah) [inline src/operations/course/CourseList.tsx]: if(isTeacher()) return ....
return ...

File: src/operations/course/CourseShow.jsx [hunk 13]
```diff
@@ -1,75 +1,112 @@
...
+          }}
+          source={source}
+        />
+      </Box>
+    </Box>
+  );
+};
+
+function CourseShow() {
+  const {id} = useParams();
```
Comment (NyAndoMayah) [inline src/operations/course/CourseShow.jsx:49]: skip a line

File: src/operations/course/CourseShow.jsx [hunk 14]
```diff
@@ -1,75 +1,112 @@
...
+import {PALETTE_COLORS} from "@/haTheme";
 
-export const CourseShowDialog = ({courseId}) => {
-  const [showDetails, toggleShowDetails] = useToggle();
+const LabeledIconField = ({
+  icon: Icon,
+  label,
+  source,
+  iconColor = "inherit",
+}) => {
```
Comment (NyAndoMayah) [inline src/operations/course/CourseShow.jsx]: => (

File: src/providers/awardedCoursesProvider.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import {teachingApi} from "@/providers/api";
+import {CreateAwardedCourse} from "@haapi/typescript-client";
```
Comment (NyAndoMayah) [inline src/providers/awardedCoursesProvider.ts:1]: reorder imports

File: src/providers/dataProvider.ts [hunk 16]
```diff
@@ -25,6 +25,7 @@ import studentLettersProvider from "@/providers/studentLettersProvider";
 import lettersProvider from "@/providers/lettersProvider";
 import lettersStatsProvider from "./letterStatsProvider";
 import receiptProvider from "./receiptProvider";
+import awardedCoursesProvider from "./awardedCoursesProvider";
```
Comment (NyAndoMayah) [inline src/providers/dataProvider.ts:28]: skip a line

**Review by NyAndoMayah — CHANGES_REQUESTED** [review]

File: cypress/e2e/receipt.cy.tsx [hunk 17]
```diff
@@ -40,8 +40,9 @@ describe("Student receipt", () => {
       force: true,
     });
     cy.getByTestid("get-receipt-btn").click();
-    cy.wait("@downloadReceipt");
-    cy.getByTestid("file-link").and("have.attr", "href").and("include", "blob");
+    cy.wait("@downloadReceipt").then((interception) => {
+      expect(interception.response!.statusCode).to.eq(200);
```
Comment (JeanMarc-RAJAONARIVELONA) [inline cypress/e2e/receipt.cy.tsx]: .its("response.status").should("req", 200)

File: cypress/e2e/receipt.cy.tsx [hunk 18]
```diff
@@ -99,7 +100,8 @@ describe("Manager receipt", () => {
     cy.get("#main-content tbody tr").first().click();
     cy.wait("@getFee1");
     cy.getByTestid("get-receipt-btn").click();
-    cy.wait("@downloadReceipt");
-    cy.getByTestid("file-link").and("have.attr", "href").and("include", "blob");
+    cy.wait("@downloadReceipt").then((interception) => {
```
Comment (JeanMarc-RAJAONARIVELONA) [inline cypress/e2e/receipt.cy.tsx]: same as previous

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+import {
+  Create,
+  required,
+  SelectInput,
+  SimpleForm,
+  useGetList,
+} from "react-admin";
+import {useParams} from "react-router-dom";
+
+interface AwardedCoursesCreateProps {
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: export!

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+  toggleShowCreate: () => void;
+}
+
+export const AwardedCoursesCreate: FC<AwardedCoursesCreateProps> = ({
+  toggleShowCreate,
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx:17]: toggleShowCreate -> onSuccess

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+}) => {
+  const {data: teachers} = useGetList("teachers");
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: {data: teachers = []}

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+  const {data: groups} = useGetList("groups");
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: {data: groups = []}

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+  const courseId = params.id;
+
+  const teacherChoices = teachers
+    ? teachers.map((teacher) => ({
+        id: teacher.id,
+        name: `${teacher.first_name} ${teacher.last_name}`,
+      }))
+    : [];
+
+  const groupChoices = groups
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: useless condition... add a default value on the destructuration and remove this useless condition

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+  toggleShowCreate,
+}) => {
+  const {data: teachers} = useGetList("teachers");
+  const {data: groups} = useGetList("groups");
+
+  const params = useParams();
+  const courseId = params.id;
+
+  const teacherChoices = teachers
+    ? teachers.map((teacher) => ({
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: put a default value on the destructuration then remove this useless condition

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    <Create
+      resource="awarded-courses"
+      title=" "
+      redirect={() => `course/${courseId}/show`}
+      mutationOptions={{
+        onSuccess: () => {
+          toggleShowCreate();
+        },
+      }}
+      transform={(data: any) => ({
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: any -> AwardedCourse

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+        ...data,
+        course_id: courseId,
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: use Create.defaultValues props to inject the courseId

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+export const AwardedCoursesCreate: FC<AwardedCoursesCreateProps> = ({
+  toggleShowCreate,
+}) => {
+  const {data: teachers} = useGetList("teachers");
+  const {data: groups} = useGetList("groups");
+
+  const params = useParams();
+  const courseId = params.id;
+
+  const teacherChoices = teachers
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: use useMemo

File: src/operations/awardedCourses/AwardedCourseCreate.tsx (hunk 23 above)
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: useMemo

File: src/operations/awardedCourses/TeacherCourseList.tsx [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+  const teacherId = isTeacher() ? authProvider.getCachedWhoami().id : null;
+
+  return (
+    <HaList
+      icon={<BookIcon />}
+      title="Cours"
+      resource="awarded-courses"
+      actions={false}
+      listProps={{
+        filterDefaultValues: {teacherId: teacherId},
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/TeacherCourseList.tsx]: use filter prop instead, as filterDefaultValues is mutable throught useListcontext()

File: src/operations/course/CourseList.tsx [hunk 29]
```diff
@@ -1,22 +1,25 @@
-import {useState} from "react";
-import {TextField, useNotify} from "react-admin";
+import {ShowButton, TextField, useNotify} from "react-admin";
 import {Box} from "@mui/material";
 import {Book, Add as AddIcon} from "@mui/icons-material";
+import {CourseListFilter} from "@/operations/course/components";
+import {CourseCreate} from "@/operations/course/CourseCreate";
+import {TeacherCourseList} from "@/operations/awardedCourses/TeacherCourseList";
+import {CourseEditButton} from "./CourseEditButton";
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/course/CourseList.tsx]: fix import path

File: src/operations/awardedCourses/TeacherCourseList.tsx [hunk 30]
```diff
@@ -0,0 +1,29 @@
+import {TextField} from "react-admin";
+import {MenuBook as BookIcon} from "@mui/icons-material";
+import {useRole} from "@/security/hooks";
+import authProvider from "@/providers/authProvider";
+import {HaList} from "@/ui/haList";
+
+export const TeacherCourseList = () => {
+  const {isTeacher} = useRole();
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/TeacherCourseList.tsx]: const teacher = authProvider.getCachedWhoami()
if (!teacher) return null;

File: src/operations/awardedCourses/AwardedCoursesList.tsx (hunk 7 above)
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCoursesList.tsx]: fix the filename

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+  SimpleForm,
+  useGetList,
+} from "react-admin";
+import {useParams} from "react-router-dom";
+
+interface AwardedCoursesCreateProps {
+  toggleShowCreate: () => void;
+}
+
+export const AwardedCoursesCreate: FC<AwardedCoursesCreateProps> = ({
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx]: AwardedCourseList or AwardedCourses

File: src/operations/awardedCourses/AwardedCourseCreate.tsx [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+  toggleShowCreate,
+}) => {
+  const {data: teachers} = useGetList("teachers");
+  const {data: groups} = useGetList("groups");
+
+  const params = useParams();
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/awardedCourses/AwardedCourseCreate.tsx:22]: const { courseId } = useParams();

if (!courseId) return null;

File: src/operations/course/CourseList.tsx [hunk 33]
```diff
@@ -1,22 +1,25 @@
...
 import {HaList} from "@/ui/haList";
 import {ButtonBase, HaActionWrapper} from "@/ui/haToolbar";
 import {Dialog} from "@/ui/components";
-import {CourseCreate} from "@/operations/course/CourseCreate";
-import {useToggle} from "@/hooks";
-import {CourseShowDialog} from "@/operations/course/CourseShow";
-import {CourseListFilter} from "@/operations/course/components";
-import {CourseEditButton} from "@/operations/course/CourseEditButton";
 
 export function CourseList() {
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/operations/course/CourseList.tsx:15]: put //TODO: refactor, this CourseLIst facade

File: src/providers/awardedCoursesProvider.ts [hunk 34]
```diff
@@ -0,0 +1,34 @@
...
+import {teachingApi} from "@/providers/api";
+import {HaDataProviderType} from "@/providers/HaDataProviderType";
+
+const awardedCoursesProvider: HaDataProviderType = {
+  getList: async (page, perPage, filter, _meta) => {
```
Comment (JeanMarc-RAJAONARIVELONA) [inline src/providers/awardedCoursesProvider.ts]: could throws npe in prod!

filter is possibly null, you didnt write any conditions that ensures that filter.courseId or filter.teacherId wont be null...

add filter = {}

(error snippet: "cannot read properties of undefined ("courseId")

File: cypress/e2e/receipt.cy.tsx [hunk 35]
```diff
@@ -40,8 +40,9 @@ describe("Student receipt", () => {
       force: true,
```
Comment (JeanMarc-RAJAONARIVELONA) [inline cypress/e2e/receipt.cy.tsx:40]: where is the Awarded course e2e test

**Review by JeanMarc-RAJAONARIVELONA — CHANGES_REQUESTED** [review]

**Pull request merged**